src/qmozenginesettings.cpp

```diff
@@ -220,24 +220,37 @@ void QMozEngineSettingsPrivate::setPreference(const QString &key, const QVariant
 
     mozilla::embedlite::EmbedLiteApp *app = QMozContext::instance()->GetApp();
 
-    int type = value.type();
-    switch (type) {
-    case QMetaType::QString:
-    case QMetaType::Float:
-    case QMetaType::Double:
+    switch (preferenceType) {
+    case QMozEngineSettings::BoolPref:
+        app->SetBoolPref(key.toUtf8().data(), value.toBool());
+        break;
+    case QMozEngineSettings::StringPref:
         app->SetCharPref(key.toUtf8().data(), value.toString().toUtf8().data());
         break;
-    case QMetaType::Int:
-    case QMetaType::UInt:
-    case QMetaType::LongLong:
-    case QMetaType::ULongLong:
+    case QMozEngineSettings::IntPref:
         app->SetIntPref(key.toUtf8().data(), value.toInt());
         break;
-    case QMetaType::Bool:
-        app->SetBoolPref(key.toUtf8().data(), value.toBool());
-        break;
     default:
-        qCWarning(lcEmbedLiteExt) << "Unknown pref" << key << "value" << value << "type:" << value.type() << "userType:" << value.userType();
+        int type = value.type();
+        switch (type) {
+        case QMetaType::QString:
+        case QMetaType::Float:
+        case QMetaType::Double:
```

Comment:
I'm wondering if it would make sense to interpret `Float` and `Double` as the `SetIntPref` type, so that the QML and C++ types align better with the gecko types. Currently the only values passed in as doubles are `embedlite.zoomMargin` and `embedlite.inputItemSize`, neither of which appear to be used (so should probably be removed).

It's just an idea, but this would also then avoid the need to change the `setPreferences` API, since this appears to be how the original QML in sailfish-browser was intended to work.

Reply:
Not good idea, as those do not fit into an integer. Original QML regarding ConfigDialog of browser has never worked correctly. There are cases like "layout.css.devPixelsPerPx" which is string (float) but can be even number like "3" => Saving it as int would fail. Similarly for instance "extensions.lastAppBuildId" is double 20211214143622 but of type string.

Reply:
Ah, I mean the other way around. Gecko "doubles" (which don't really exist) should always be passed to `setPreferences` as `QVariant::String` types which would allow gecko integer types to then be passed as `QVariant::Double`, since this is what QML does.

What I'm suggesting would only really affect the C++, since it could then be potentially confusing if someone passes a double and expects it to be stored as a string. That's not happening anywhere in practice.

But, the current changes are working fine; I don't have any objection to them going in as they are.

As an aside, it's frustrating that QML has an integer type but Javascript doesn't.